File: frontend/src/lib/api/study-plans.ts
/**
 * Study Plan API client.
 */

import { apiClient } from './client';
import type {
  StudyPlan,
  StudyPlanCreate,
  StudyPlanUpdate,
  StudyPlanFull,
  StudyPlanStats,
  StudyTopic,
  StudyTopicCreate,
  StudyTopicUpdate,
  StudySession,
  StudySessionCreate,
  StudySessionUpdate,
  PlanAdjustment,
  WeeklySchedule,
  PlanGenerationRequest,
  PlanGenerationResult,
} from '../types/study-plan';

// ============ Study Plan API ============

export async function createStudyPlan(data: StudyPlanCreate): Promise<StudyPlan> {
  const response = await apiClient.post<StudyPlan>('/study-plans', data);
  return response.data;
}

export async function generateStudyPlan(request: PlanGenerationRequest): Promise<PlanGenerationResult> {
  const response = await apiClient.post<PlanGenerationResult>('/study-plans/generate', request);
  return response.data;
}

export async function getStudyPlans(notebookId?: string, activeOnly = false): Promise<StudyPlan[]> {
  const params = new URLSearchParams();
  if (notebookId) params.append('notebook_id', notebookId);
  if (activeOnly) params.append('active_only', 'true');

  const response = await apiClient.get<StudyPlan[]>(`/study-plans?${params.toString()}`);
  return response.data;
}

export async function getTodaySessions(planId?: string): Promise<StudySession[]> {
  const params = planId ? `?plan_id=${encodeURIComponent(planId)}` : '';
  const response = await apiClient.get<StudySession[]>(`/study-plans/today${params}`);
  return response.data;
}

export async function getStudyPlan(planId: string): Promise<StudyPlanFull> {
  const response = await apiClient.get<StudyPlanFull>(`/study-plans/${encodeURIComponent(planId)}`);
  return response.data;
}

export async function updateStudyPlan(planId: string, data: StudyPlanUpdate): Promise<StudyPlan> {
  const response = await apiClient.patch<StudyPlan>(`/study-plans/${encodeURIComponent(planId)}`, data);
  return response.data;
}

export async function deleteStudyPlan(planId: string): Promise<void> {
  await apiClient.delete(`/study-plans/${encodeURIComponent(planId)}`);
}

export async function getStudyPlanStats(planId: string): Promise<StudyPlanStats> {
  const response = await apiClient.get<StudyPlanStats>(`/study-plans/${encodeURIComponent(planId)}/stats`);
  return response.data;
}

export async function getWeeklySchedule(planId: string, weekStart?: string): Promise<WeeklySchedule> {
  const params = weekStart ? `?week_start=${weekStart}` : '';
  const response = await apiClient.get<WeeklySchedule>(`/study-plans/${encodeURIComponent(planId)}/schedule${params}`);
  return response.data;
}

// ============ Topic API ============

export async function createTopic(planId: string, data: StudyTopicCreate): Promise<StudyTopic> {
  const response = await apiClient.post<StudyTopic>(`/study-plans/${encodeURIComponent(planId)}/topics`, data);
  return response.data;
}

export async function getTopics(planId: string): Promise<StudyTopic[]> {
  const response = await apiClient.get<StudyTopic[]>(`/study-plans/${encodeURIComponent(planId)}/topics`);
  return response.data;
}

export async function getTopic(topicId: string): Promise<StudyTopic> {
  const response = await apiClient.get<StudyTopic>(`/study-plans/topics/${encodeURIComponent(topicId)}`);
  return response.data;
}

export async function updateTopic(topicId: string, data: StudyTopicUpdate): Promise<StudyTopic> {
  const response = await apiClient.patch<StudyTopic>(`/study-plans/topics/${encodeURIComponent(topicId)}`, data);
  return response.data;
}

export async function deleteTopic(topicId: string): Promise<void> {
  await apiClient.delete(`/study-plans/topics/${encodeURIComponent(topicId)}`);
}

// ============ Session API ============

export async function createSession(planId: string, data: StudySessionCreate): Promise<StudySession> {
  const response = await apiClient.post<StudySession>(`/study-plans/${encodeURIComponent(planId)}/sessions`, data);
  return response.data;
}

export async function getSessions(planId: string): Promise<StudySession[]> {
  const response = await apiClient.get<StudySession[]>(`/study-plans/${encodeURIComponent(planId)}/sessions`);
  return response.data;
}

export async function getSession(sessionId: string): Promise<StudySession> {
  const response = await apiClient.get<StudySession>(`/study-plans/sessions/${encodeURIComponent(sessionId)}`);
  return response.data;
}

export async function updateSession(sessionId: string, data: StudySessionUpdate): Promise<StudySession> {
  const response = await apiClient.patch<StudySession>(`/study-plans/sessions/${encodeURIComponent(sessionId)}`, data);
  return response.data;
}

export async function startSession(sessionId: string): Promise<StudySession> {
  const response = await apiClient.post<StudySession>(`/study-plans/sessions/${encodeURIComponent(sessionId)}/start`);
  return response.data;
}

export async function completeSession(
  sessionId: string,
  rating?: number,
  notes?: string
): Promise<StudySession> {
  const params = new URLSearchParams();
  if (rating !== undefined && rating !== null) params.append('rating', rating.toString());
  if (notes) params.append('notes', notes);

  const response = await apiClient.post<StudySession>(
    `/study-plans/sessions/${encodeURIComponent(sessionId)}/complete?${params.toString()}`
  );
  return response.data;
}

export async function skipSession(sessionId: string, reason?: string): Promise<StudySession> {
  const params = reason ? `?reason=${encodeURIComponent(reason)}` : '';
  const response = await apiClient.post<StudySession>(
    `/study-plans/sessions/${encodeURIComponent(sessionId)}/skip${params}`
  );
  return response.data;
}

export async function deleteSession(sessionId: string): Promise<void> {
  await apiClient.delete(`/study-plans/sessions/${encodeURIComponent(sessionId)}`);
}

// ============ Adjustment API ============

export async function getAdjustments(planId: string): Promise<PlanAdjustment[]> {
  const response = await apiClient.get<PlanAdjustment[]>(`/study-plans/${encodeURIComponent(planId)}/adjustments`);
  return response.data;
}

export async function respondToAdjustment(adjustmentId: string, accepted: boolean): Promise<void> {
  await apiClient.post(`/study-plans/adjustments/${encodeURIComponent(adjustmentId)}/respond`, {
    adjustment_id: adjustmentId,
    accepted,
  });
}

File: api/routers/study_plans.py
"""
Study Plan API Router.
AI-generated personalized study schedules.
"""

from datetime import datetime
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Query
from loguru import logger

from open_notebook.domain.study_plan import (
    StudyPlan, StudyPlanCreate, StudyPlanUpdate,
    StudyPlanFull, StudyPlanWithTopics, StudyPlanStats,
    StudyTopic, StudyTopicCreate, StudyTopicUpdate,
    StudySession, StudySessionCreate, StudySessionUpdate,
    PlanAdjustment, PlanAdjustmentResponse,
    WeeklySchedule, DailySchedule,
    PlanGenerationRequest, PlanGenerationResult
)
from open_notebook.services.study_plan_service import study_plan_service

router = APIRouter(prefix="/study-plans", tags=["study-plans"])


# ============ Study Plan Endpoints ============

@router.post("", response_model=StudyPlan)
async def create_plan(data: StudyPlanCreate):
    """Create a new study plan."""
    return await study_plan_service.create_plan(data)


@router.post("/generate", response_model=PlanGenerationResult)
async def generate_plan(request: PlanGenerationRequest):
    """Generate a complete study plan using AI."""
    return await study_plan_service.generate_plan(request)


@router.get("", response_model=List[StudyPlan])
async def list_plans(
    notebook_id: Optional[str] = Query(None, description="Filter by notebook ID"),
    active_only: bool = Query(False, description="Only return active plans")
):
    """List study plans."""
    if notebook_id:
        plans = await study_plan_service.get_plans_for_notebook(notebook_id)
    elif active_only:
        plans = await study_plan_service.get_active_plans()
    else:
        plans = await study_plan_service.get_active_plans()  # Default to active
    return plans


@router.get("/today", response_model=List[StudySession])
async def get_today_sessions(plan_id: Optional[str] = Query(None)):
    """Get study sessions scheduled for today."""
    return await study_plan_service.get_today_sessions(plan_id)


@router.get("/{plan_id}", response_model=StudyPlanFull)
async def get_plan(plan_id: str):
    """Get a study plan with all details."""
    plan = await study_plan_service.get_plan_full(plan_id)
    if not plan:
        raise HTTPException(status_code=404, detail="Study plan not found")
    return plan


@router.patch("/{plan_id}", response_model=StudyPlan)
async def update_plan(plan_id: str, data: StudyPlanUpdate):
    """Update a study plan."""
    plan = await study_plan_service.update_plan(plan_id, data)
    if not plan:
        raise HTTPException(status_code=404, detail="Study plan not found")
    return plan


@router.delete("/{plan_id}")
async def delete_plan(plan_id: str):
    """Delete a study plan and all related data."""
    success = await study_plan_service.delete_plan(plan_id)
    if not success:
        raise HTTPException(status_code=404, detail="Study plan not found")
    return {"status": "deleted", "plan_id": plan_id}


@router.get("/{plan_id}/stats", response_model=StudyPlanStats)
async def get_plan_stats(plan_id: str):
    """Get statistics for a study plan."""
    return await study_plan_service.get_plan_stats(plan_id)


@router.get("/{plan_id}/schedule", response_model=WeeklySchedule)
async def get_weekly_schedule(
    plan_id: str,
    week_start: Optional[datetime] = Query(None, description="Start of week (defaults to current week)")
):
    """Get weekly schedule for a study plan."""
    return await study_plan_service.get_weekly_schedule(plan_id, week_start)


# ============ Topic Endpoints ============

@router.post("/{plan_id}/topics", response_model=StudyTopic)
async def create_topic(plan_id: str, data: StudyTopicCreate):
    """Create a study topic."""
    if data.plan_id != plan_id:
        data.plan_id = plan_id
    return await study_plan_service.create_topic(data)


@router.get("/{plan_id}/topics", response_model=List[StudyTopic])
async def list_topics(plan_id: str):
    """List topics for a study plan."""
    return await study_plan_service.get_topics_for_plan(plan_id)


@router.get("/topics/{topic_id}", response_model=StudyTopic)
async def get_topic(topic_id: str):
    """Get a study topic."""
    topic = await study_plan_service.get_topic(topic_id)
    if not topic:
        raise HTTPException(status_code=404, detail="Topic not found")
    return topic


@router.patch("/topics/{topic_id}", response_model=StudyTopic)
async def update_topic(topic_id: str, data: StudyTopicUpdate):
    """Update a study topic."""
    topic = await study_plan_service.update_topic(topic_id, data)
    if not topic:
        raise HTTPException(status_code=404, detail="Topic not found")
    return topic


@router.delete("/topics/{topic_id}")
async def delete_topic(topic_id: str):
    """Delete a study topic."""
    success = await study_plan_service.delete_topic(topic_id)
    if not success:
        raise HTTPException(status_code=404, detail="Topic not found")
    return {"status": "deleted", "topic_id": topic_id}


# ============ Session Endpoints ============

@router.post("/{plan_id}/sessions", response_model=StudySession)
async def create_session(plan_id: str, data: StudySessionCreate):
    """Create a study session."""
    if data.plan_id != plan_id:
        data.plan_id = plan_id
    return await study_plan_service.create_session(data)


@router.get("/{plan_id}/sessions", response_model=List[StudySession])
async def list_sessions(plan_id: str):
    """List sessions for a study plan."""
    return await study_plan_service.get_sessions_for_plan(plan_id)


@router.get("/sessions/{session_id}", response_model=StudySession)
async def get_session(session_id: str):
    """Get a study session."""
    session = await study_plan_service.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


@router.patch("/sessions/{session_id}", response_model=StudySession)
async def update_session(session_id: str, data: StudySessionUpdate):
    """Update a study session."""
    session = await study_plan_service.update_session(session_id, data)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


@router.post("/sessions/{session_id}/start", response_model=StudySession)
async def start_session(session_id: str):
    """Start a study session."""
    session = await study_plan_service.start_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


@router.post("/sessions/{session_id}/complete", response_model=StudySession)
async def complete_session(
    session_id: str,
    rating: Optional[int] = Query(None, ge=1, le=5),
    notes: Optional[str] = Query(None)
):
    """Complete a study session."""
    try:
        logger.info(f"API: complete_session called with session_id={session_id}, rating={rating}, notes={notes}")
        session = await study_plan_service.complete_session(session_id, rating, notes)
        if not session:
            raise HTTPException(status_code=404, detail="Session not found")
        logger.info(f"API: Session completed successfully: {session.id}")
        return session
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"API: Error completing session: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/sessions/{session_id}")
async def delete_session(session_id: str):
    """Delete a study session."""
    success = await study_plan_service.delete_session(session_id)
    if not success:
        raise HTTPException(status_code=404, detail="Session not found")
    return {"status": "deleted", "session_id": session_id}


@router.post("/sessions/{session_id}/skip", response_model=StudySession)
async def skip_session(
    session_id: str,
    reason: Optional[str] = Query(None, description="Reason for skipping")
):
    """Skip a study session."""
    session = await study_plan_service.skip_session(session_id, reason)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


# ============ Adjustment Endpoints ============

@router.get("/{plan_id}/adjustments", response_model=List[PlanAdjustment])
async def list_adjustments(plan_id: str):
    """List adjustments for a study plan."""
    return await study_plan_service.get_adjustments_for_plan(plan_id)


@router.post("/adjustments/{adjustment_id}/respond")
async def respond_to_adjustment(adjustment_id: str, response: PlanAdjustmentResponse):
    """Accept or reject a plan adjustment."""
    success = await study_plan_service.respond_to_adjustment(adjustment_id, response.accepted)
    return {"status": "accepted" if response.accepted else "rejected", "adjustment_id": adjustment_id}
